docs/architecture/adr-11-refactor-errors.md

```diff
@@ -0,0 +1,348 @@
+error handling architecture should strive to achieve the following goals:
+
+1. Empower host developers to respond to ibc-rs-originating errors that affect host logic.
+2. Assist ibc-rs core developers' debugging efforts by providing detailed, human-readable
+   error reports and/or stack traces.
+
+These two aims highlight the need to clearly distinguish between *host*-level errors and
+*protocol*-level errors. Developers working at the host level often do not need to concern
+themselves with lower-level protocol errors; even if they do, those errors should not be
+exposed to host developers at the fine-grained level of granularity that is more appropriate
+for protocol-level developers.
+
+As it currently stands, ibc-rs exposes a top-level `ContextError` type that is returned
+by all validation and execution context methods, which are those that need to be implemented
+by hosts as part of the ibc-rs integration process. This `ContextError` type encapsulates all
+of the protocol-level errors that ibc-rs exposes. As a result, the onus is being placed on
+host developers to decide whether these errors are relevant or not. In other words, the
+distinction between host- and protocol-level errors is not reflected in ibc-rs's error
+handling methodology.
+
+## Proposal
+
+### Some Changes to `ContextError`
+
+ibc-rs's top-level error type, `ContextError`, will be renamed to `HandlerError` in an effort
+to improve the semantics of this error type. `HandlerError` will be returned only by the top-
+level handler entrypoints of ibc-rs, namely the [dispatch], [validate], and [execute] methods.
+
+```diff
+- pub enum ContextError {
++ pub enum HandlerError {
+    /// ICS02 Client error: {0}
+    Client(ClientError),
+    /// ICS03 Connection error: {0}
+    Connection(ConnectionError),
+    /// ICS04 Channel error: {0}
+    Channel(ChannelError),
+-   /// ICS04 Packet error: {0}
+-   Packet(PacketError),
+    /// ICS26 Router error: {0}
+    Router(RouterError),
+}
+```
+
+In addition, the error variant that contained the `PacketError` type will be removed. This
+is because we're opting to move towards only having a single error type for each ICS module,
+removing the discrepancy that the ICS04 module exposes two distinct error types. The
+`PacketError` type will be merged into the `ChannelError` type.
+
+### The New `HostError` Type
+
+In light of the stated rationale of making it clear when an error originates from host logic
+vs ibc-rs's internal logic, we propose adding a new `HostError` type that will live as a
+variant in each of the module-level error types, as well as in the application-level error
+types, `TokenTransferError` and `NftTransferError`. Initially, we had only a single 
+`Host(HostError)` variant that existed in the `HandlerError`type, but it became clear 
+that this wasn't the correct place in which to expose host-level errors. This is because 
```

Comment:
```suggestion
types, `TokenTransferError` and `NftTransferError`. Initially, we had only a single
`Host(HostError)` variant that existed in the `HandlerError`type, but it became clear
that this wasn't the correct place in which to expose host-level errors. This is because
```